### DISCUSSION TITLE
Refoss-P11: Temperature value jumps back and forth when switching ON/OFF

### DISCUSSION THREAD

**Discussion by broth-itk** [arendst/Tasmota#19327] [General]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._

Tasmota 13.1.0 on Refoss-P11: Internal Temperature makes jumps +- 1°C when switching ON/OFF

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): Refoss-P11
- [x] Tasmota binary firmware version number used: Preinstalled, updates to latest version
  - [x] Pre-compiled
  - [ ] Self-compiled
- [ ] Flashing tools used: _____
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:

11:12:05.449 CMD: Backlog Template; Module; GPIO 255
11:12:05.471 RSL: RESULT = {"NAME":"Refoss-P11","GPIO":[0,0,0,3104,576,32,0,0,320,224,0,0,0,4736],"FLAG":0,"BASE":1}
11:12:05.673 RSL: RESULT = {"Module":{"0":"Refoss-P11"}}
11:12:05.881 RSL: RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"3104":"CSE7766 Rx"},"GPIO4":{"576":"LedLink_i"},"GPIO5":{"32":"Button1"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"320":"Led_i1"},"GPIO13":{"224":"Relay1"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"4736":"ADC Temp"}}

```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:

11:12:35.435 CMD: status 0
11:12:35.442 RSL: STATUS = {"Status":{"Module":0,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"tasmota_91E3E2","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
11:12:35.448 RSL: STATUS1 = {"StatusPRM":{"Baudrate":4800,"SerialConfig":"8E1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota-4M.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:10:13","StartupUTC":"2023-08-16T10:02:22","Sleep":50,"CfgHolder":4617,"BootCount":8,"BCResetTime":"2023-08-14T11:39:38","SaveCount":57,"SaveAddress":"F8000"}}
11:12:35.452 RSL: STATUS2 = {"StatusFWR":{"Version":"13.1.0(tasmota-4M)","BuildDateTime":"2023-08-15T07:58:57","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"416/699"}}
11:12:35.457 RSL: STATUS3 = {"StatusLOG":{"SerialLog":0,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["Umbrella IoT",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000680003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
11:12:35.469 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":706,"Free":1340,"Heap":21,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"164068","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","8F9AC78F","04368001","000000CF","010013C0","C000F981","00004084","00201000","54000020","00000080"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,38,45,50,62,68","Sensors":"1,2,3,4,5,6","I2CDriver":"7"}}
11:12:35.476 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota-91E3E2-0994","IPAddress":"192.168.202.102","Gateway":"192.168.202.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.202.1","DNSServer2":"0.0.0.0","Mac":"34:94:54:91:E3:E2","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
11:12:35.481 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_91E3E2","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
11:12:35.488 RSL: STATUS7 = {"StatusTIM":{"UTC":"2023-08-16T10:12:35","Local":"2023-08-16T11:12:35","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"05:44","Sunset":"20:04"}}
11:12:35.491 RSL: STATUS9 = {"StatusPTH":{"PowerDelta":[0,0,0],"PowerLow":0,"PowerHigh":0,"VoltageLow":0,"VoltageHigh":0,"CurrentLow":0,"CurrentHigh":0}}
11:12:35.499 RSL: STATUS10 = {"StatusSNS":{"Time":"2023-08-16T11:12:35","ANALOG":{"Temperature":27.0},"ENERGY":{"TotalStartTime":"2023-08-14T11:39:38","Total":0.000,"Yesterday":0.000,"Today":0.000,"Power":0,"ApparentPower":0,"ReactivePower":0,"Factor":0.00,"Voltage":0,"Current":0.000},"TempUnit":"C"}}
11:12:35.505 RSL: STATUS11 = {"StatusSTS":{"Time":"2023-08-16T11:12:35","Uptime":"0T00:10:13","UptimeSec":613,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"Umbrella IoT","BSSId":"E8:ED:D6:88:B4:31","Channel":6,"Mode":"11n","RSSI":76,"Signal":-62,"LinkCount":1,"Downtime":"0T00:00:04"}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

TOGGLE ON/OFF and watch temperature value carefully.

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

Temperature measurement should be stable.
The issue is independent of load. It happens with and without load.

Possibly the ADC is experiencing interference due to the switching of relay coil.
Or maybe the additional load of the coil has an effect on the ADCs reference voltage.

As a hardware fix is not practicable, I propose to generate a moving mean value and show that on the web interface.
Temperature changes are slow and this should not cause any sideeffects.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

n/a

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

n/a

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by arendst** [comment]
WARNING: Set ``TempRes 1`` and watch the horrific temp changes at every GUI update.

The temp sensor is INSIDE the power plug measuring internal temperature raised and lowered by ESP8266 activity and mostly wifi connection power usage.

Do not expect a stable temperature as the internals is heated by extra power needed for the relay too.

In short, this is not an issue but a result of internal temperature sensor location. I'll move this issue to discussions for anyone wanting to shine a different light over this.

**Comment by broth-itk** [comment]
@arendst Thank you! This did actually fix the issue.

Would it be possible to add this setting to the supported device information page?
https://templates.blakadder.com/refoss_P11.html
(if managed by you or an associated party)

Like the ADC calibration to fix invalid temperature readings.

**Reply by evidalr** [reply]
**broth-itk** said: _"I propose to generate a moving mean value and show that on the web interface. Temperature changes are slow and this should not cause any sideeffects."_

The following ruleset computes a moving average (kind of) and reports the result on the web UI via Counter1 (which I have configured on GPIO-14):
```
Rule2
  ON analog#Temperature DO IF (var6==1) backlog
     var6 0 ; var5=(var4*var5+10*%value%)/(var4+1) ; add4 1 ; delay 10 ; var6 1
  ENDIF ENDON
  ON var4#state|4 DO counter1 %var5% ENDON
  ON var4#state==32 DO var4 16 ENDON 

backlog var4 0 ; var5 0 ; var6 1 ; rule2 1
```
This stabilises the originally somewhat errating temperature readings.  But it does NOT solve the dependence of the reading on the status of the relay, which is caused by a real temperature increase of a few degrees inside the device because of the heat produced by the relay.

A couple of rules can be easily added to the above ruleset to recalibrate the ADC parameters when Power1#state=1 in order to compensate for the extra heat, and reset the normal calibration when Power1#state=0.

But to do this properly we would need to know what is the function of each of the adcParam values, which in my case are: [17,4000,80000,4250,10000].  Any one have an idea?

**Reply by sfromis** [reply]
The function of your AdcParam values depend on what ADC pin type you configured.
Meanings are explained here: https://tasmota.github.io/docs/ADC/#commands
(The 17 is simply the pin number for the analog pin, does not indicate your config)

**Reply by evidalr** [reply]
Thanks.  Yes I had seen this, but not sure to understand it all well enough:
```
2 = Temperature Steinhart-Hart thermistor equation parameters:
<param1> = NTC Voltage bridge resistor in Ohms (default = 32000)
<param2> = NTC Resistance in Ohms (default = 10000)
<param3> = NTC Beta Coefficient (default = 3350)
```
First, I miss the forth parameter (10000 in my case).  Second, I guess the beta coefficient is the one denoted with `B` in the formula 1/T = 1/T0 + 1/B ln R/R0 in this link: (https://en.wikipedia.org/wiki/Thermistor), reached from the Steinhaart-Hart equation link (https://en.wikipedia.org/wiki/Steinhart%E2%80%93Hart_equation), found in (https://tasmota.github.io/docs/ADC/#commands) -- but I'm not sure how these two resitors R, R0 are related with  param1 and param2 ... 

It would be nice to have an idea of the typical circuit connecting the NTC thermistor to the analog pin. I will dig further abouth this as time permits.

**Reply by evidalr** [reply]
Well, I have found these two blogs which surely provide all what is needed:
(https://learn.adafruit.com/thermistor/using-a-thermistor) and (https://arduino.stackexchange.com/questions/36505/help-using-thermistor-with-esp8266).
Now I need time to read them in detail.

**Reply by evidalr** [reply]
No need of detailed calibration.  Just with TempOffset the Refoss-P11 temperature measurements can be easily adjusted to achieve room temperature readings which typically only differ from those of a reference thermometer in +- 0.5 ºC, with occasional differences of at most +- 1ºC.  I have two of these devices and in both I had to set the TempOffset parameter to the same value: 1.0.

Of course, this is with Power1 (the relay) in the OFF state and using the moving average rule set I posted in my message of Mar 8.

To re-calibrate the readings for Power1 ON, I have added the following two rules:
```
Rule2 +
  ON Power1#state=1 DO TempOffset -5.0 ENDON
  ON Power1#state=0 DO TempOffset  1.0 ENDON
```
Then, readings get significantly off (up to +-6 ºC) for a while.  But as the device internal temperature becomes stable, the readings again go progressively close to the reference room temperature, with a limit typical error of +-0.5 ºC and occasional jumps up to +-1 ºC.

The bad-readings period after an OFF to ON transition lasts about 10-15 minutes, and about 15--20 minutes when the relay goes from ON to OFF.

While not an ideal behaviour, this makes the temperature readings of this device useful enough for many use cases.

**Comment by bam80** [comment]
Very interesting.
I'm considering this device for the purchase as I didn't find anything Tasmota-compatible with external sensor and ability to measure power:
https://github.com/arendst/Tasmota/discussions/20125

@broth-itk would you share your experience of using the device, especially interesting how internal temp sensor's data is accurate under load and without (comparing with ambient temperature).

**Reply by sfromis** [reply]
As already explained in [the other thread](https://github.com/arendst/Tasmota/discussions/20125), do not rely on an internal sensor in a plug-type device if you want room temperature. You cannot even expect a stable difference. For ambient temperature, you need good airflow and distance.

**Reply by bam80** [reply]
I know, just wanted to get real user experience.
I also can assume it's temp data is precise enough to organize thermostat, at least when the plug is not under load (I have some ideas for work-around of the plug self-heating).

**Comment by broth-itk** [comment]
@bam80: I agree, the internal sensor can be used only for monitoring the plug itself or to give rough indications.
There is a considerable offset and delay for external temperature changes to get recognized by the sensor.
